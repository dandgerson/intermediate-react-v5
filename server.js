import { readFileSync } from "fs";
import path from "path";
import { fileURLToPath } from "url";
import express from "express";
import { render } from "./dist/server/App.server.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PORT = process.env.PORT || 3001;

const html = readFileSync(
  path.resolve(__dirname, "./dist/client/index.html")
).toString();

const parts = html.split("not rendered");

const app = express();
app.use(
  "/assets",
  express.static(path.resolve(__dirname, "./dist/client/assets"))
);

app.use((req, res) => {
  res.write(parts[0]);

  const stream = render(req.url, {
    onShellReady() {
      // if it is the crawler, do nothing here
      stream.pipe(res);
    },
    onShellError() {
      // do Error handling here
    },
    onAllReady() {
      // if it is the crawler
      // stream.pipe(res);

      // last thing to write
      res.write(parts[1]);
      res.end();
    },
    onError(err) {
      console.error(err);
    },
  });
});

console.log(`listening on http://localhost:${PORT}`);
app.listen(PORT);
