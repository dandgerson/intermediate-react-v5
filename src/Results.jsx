import { Pet } from "./Pet";

export const Results = ({ pets }) => (
  <div className="grid grid-cols-1 gap-4 sm:grid-cols-2 lg:grid-cols-3">
    {pets.length ? (
      pets.map((pet) => (
        <Pet
          key={pet.id}
          id={pet.id}
          name={pet.name}
          animal={pet.animal}
          breed={pet.breed}
          images={pet.images}
          location={`${pet.city}, ${pet.state}`}
        />
      ))
    ) : (
      <h1>No Pets found!</h1>
    )}
  </div>
);
