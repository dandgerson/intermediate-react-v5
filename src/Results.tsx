import { Pet } from "./Pet";
import { Pet as PetType } from "./apiResponseTypes";

export const Results = ({ pets }: { pets: PetType[] }) => (
  <div className="search">
    {pets.length ? (
      pets.map((pet) => (
        <Pet
          key={pet.id}
          id={pet.id}
          name={pet.name}
          animal={pet.animal}
          breed={pet.breed}
          images={pet.images}
          location={`${pet.city}, ${pet.state}`}
        />
      ))
    ) : (
      <h1>No Pets found!</h1>
    )}
  </div>
);
