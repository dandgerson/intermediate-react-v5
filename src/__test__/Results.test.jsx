import { test, expect } from "vitest";
import { render } from "@testing-library/react";
import { StaticRouter } from "react-router-dom/server";

import { Results } from "../Results";

test("render corrects with no pets", async () => {
  const { asFragment } = render(<Results pets={[]} />);
  expect(asFragment()).toMatchSnapshot();
});

test("render corrects with pets", async () => {
  const pets = [
    {
      id: 1,
      name: "Luna",
      animal: "dog",
      city: "Seattle",
      state: "WA",
      description:
        "Luna is actually the most adorable dog in the world. Her hobbies include yelling at squirrels, aggressively napping on her owners' laps, and asking to be fed two hours before IT'S DAMN WELL TIME LUNA. Luna is beloved by her puppy parents and lazily resides currently in Seattle, Washington.",
      breed: "Havanese",
      images: [
        "http://pets-images.dev-apis.com/pets/dog25.jpg",
        "http://pets-images.dev-apis.com/pets/dog26.jpg",
        "http://pets-images.dev-apis.com/pets/dog27.jpg",
        "http://pets-images.dev-apis.com/pets/dog28.jpg",
        "http://pets-images.dev-apis.com/pets/dog29.jpg",
      ],
    },
    {
      id: 2,
      name: "Bunnahabhain",
      animal: "dog",
      city: "Minneapolis",
      state: "MN",
      description:
        "Maecenas leo odio, condimentum id, luctus nec, molestie sed, justo. Pellentesque viverra pede ac diam. Cras pellentesque volutpat dui.\n\nMaecenas tristique, est et tempus semper, est quam pharetra magna, ac consequat metus sapien ut nunc. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Mauris viverra diam vitae quam. Suspendisse potenti.",
      breed: "Goldendoodle",
      images: [
        "http://pets-images.dev-apis.com/pets/dog31.jpg",
        "http://pets-images.dev-apis.com/pets/dog30.jpg",
        "http://pets-images.dev-apis.com/pets/dog32.jpg",
        "http://pets-images.dev-apis.com/pets/dog33.jpg",
      ],
    },
    {
      id: 3,
      name: "Olive",
      animal: "dog",
      city: "Minneapolis",
      state: "MN",
      description:
        "Duis consequat dui nec nisi volutpat eleifend. Donec ut dolor. Morbi vel lectus in quam fringilla rhoncus.\n\nMauris enim leo, rhoncus sed, vestibulum sit amet, cursus id, turpis. Integer aliquet, massa id lobortis convallis, tortor risus dapibus augue, vel accumsan tellus nisi eu orci. Mauris lacinia sapien quis libero.\n\nNullam sit amet turpis elementum ligula vehicula consequat. Morbi a ipsum. Integer a nibh.",
      breed: "Boxer",
      images: ["http://pets-images.dev-apis.com/pets/dog34.jpg"],
    },
    {
      id: 4,
      name: "Sudo",
      animal: "dog",
      city: "Denver",
      state: "CO",
      description:
        "Mauris enim leo, rhoncus sed, vestibulum sit amet, cursus id, turpis. Integer aliquet, massa id lobortis convallis, tortor risus dapibus augue, vel accumsan tellus nisi eu orci. Mauris lacinia sapien quis libero.\n\nNullam sit amet turpis elementum ligula vehicula consequat. Morbi a ipsum. Integer a nibh.\n\nIn quis justo. Maecenas rhoncus aliquam lacus. Morbi quis tortor id nulla ultrices aliquet.",
      breed: "Wheaten Terrier",
      images: [
        "http://pets-images.dev-apis.com/pets/dog37.jpg",
        "http://pets-images.dev-apis.com/pets/dog38.jpg",
        "http://pets-images.dev-apis.com/pets/dog39.jpg",
      ],
    },
  ];

  const { asFragment } = render(
    <StaticRouter>
      <Results pets={pets} />
    </StaticRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});
