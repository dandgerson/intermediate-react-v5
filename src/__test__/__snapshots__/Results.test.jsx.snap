// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render corrects with no pets 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <h1>
      No Pets found!
    </h1>
  </div>
</DocumentFragment>
`;

exports[`render corrects with pets 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <a
      class="pet"
      href="/details/1"
    >
      <div
        class="image-container"
      >
        <img
          alt="Luna"
          data-testid="thumbnail"
          src="http://pets-images.dev-apis.com/pets/dog25.jpg"
        />
      </div>
      <div
        class="info"
      >
        <h1>
          Luna
        </h1>
        <h2>
          dog - Havanese - Seattle, WA
        </h2>
      </div>
    </a>
    <a
      class="pet"
      href="/details/2"
    >
      <div
        class="image-container"
      >
        <img
          alt="Bunnahabhain"
          data-testid="thumbnail"
          src="http://pets-images.dev-apis.com/pets/dog31.jpg"
        />
      </div>
      <div
        class="info"
      >
        <h1>
          Bunnahabhain
        </h1>
        <h2>
          dog - Goldendoodle - Minneapolis, MN
        </h2>
      </div>
    </a>
    <a
      class="pet"
      href="/details/3"
    >
      <div
        class="image-container"
      >
        <img
          alt="Olive"
          data-testid="thumbnail"
          src="http://pets-images.dev-apis.com/pets/dog34.jpg"
        />
      </div>
      <div
        class="info"
      >
        <h1>
          Olive
        </h1>
        <h2>
          dog - Boxer - Minneapolis, MN
        </h2>
      </div>
    </a>
    <a
      class="pet"
      href="/details/4"
    >
      <div
        class="image-container"
      >
        <img
          alt="Sudo"
          data-testid="thumbnail"
          src="http://pets-images.dev-apis.com/pets/dog37.jpg"
        />
      </div>
      <div
        class="info"
      >
        <h1>
          Sudo
        </h1>
        <h2>
          dog - Wheaten Terrier - Denver, CO
        </h2>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
